File: flutter_app/lib/sheets/sheet_library_store.dart
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:uuid/uuid.dart';

import 'sheet_entry.dart';
import 'sheet_parsed_models.dart';

/// 读写 `sheets_index.json` 与 `sheets/` 目录下的谱子页文件副本。
class SheetLibraryStore {
  static const _indexFileName = 'sheets_index.json';
  static const _sheetsSubdir = 'sheets';

  Future<Directory> _docsDir() async {
    final d = await getApplicationDocumentsDirectory();
    return d;
  }

  Future<File> _indexFile() async {
    final root = await _docsDir();
    return File(p.join(root.path, _indexFileName));
  }

  Future<Directory> _sheetsDir() async {
    final root = await _docsDir();
    final dir = Directory(p.join(root.path, _sheetsSubdir));
    if (!await dir.exists()) {
      await dir.create(recursive: true);
    }
    return dir;
  }

  Future<File> _parsedFile(String sheetId) async {
    final dir = await _sheetsDir();
    return File(p.join(dir.path, '${sheetId}_parsed.json'));
  }

  /// 读取全部条目（文件损坏时返回空列表）。
  Future<List<SheetEntry>> loadAll() async {
    final f = await _indexFile();
    if (!await f.exists()) return [];
    try {
      final text = await f.readAsString();
      final decoded = jsonDecode(text);
      if (decoded is! List) return [];
      final out = <SheetEntry>[];
      for (final x in decoded) {
        final e = SheetEntry.fromJson(x);
        if (e != null) out.add(e);
      }
      out.sort((a, b) => b.addedAtMs.compareTo(a.addedAtMs));
      return out;
    } catch (_) {
      return [];
    }
  }

  Future<void> _writeAll(List<SheetEntry> entries) async {
    final f = await _indexFile();
    final encoded = jsonEncode(entries.map((e) => e.toJson()).toList());
    await f.writeAsString(encoded);
  }

  /// 将 [sources] 按顺序复制到沙箱并追加一条谱子索引（至少一页）。
  Future<SheetEntry> importSheetPages({
    required List<File> sources,
    required String displayName,
  }) async {
    if (sources.isEmpty) {
      throw ArgumentError('sources must not be empty');
    }
    final dir = await _sheetsDir();
    final id = const Uuid().v4();
    final storedNames = <String>[];
    for (var i = 0; i < sources.length; i++) {
      var ext = p.extension(sources[i].path).toLowerCase();
      if (ext.isEmpty || ext == '.') {
        ext = '.jpg';
      }
      final name = '${id}_$i$ext';
      final dest = File(p.join(dir.path, name));
      await sources[i].copy(dest.path);
      storedNames.add(name);
    }
    final entry = SheetEntry(
      id: id,
      displayName: displayName,
      storedFileNames: storedNames,
      addedAtMs: DateTime.now().millisecondsSinceEpoch,
      parseStatus: SheetParseStatus.rawOnly.name,
    );
    final all = await loadAll();
    all.add(entry);
    await _writeAll(all);
    return entry;
  }

  /// 删除索引项及沙箱内对应页文件（若存在）。
  Future<void> remove(String id) async {
    final all = await loadAll();
    final keep = <SheetEntry>[];
    SheetEntry? victim;
    for (final e in all) {
      if (e.id == id) {
        victim = e;
      } else {
        keep.add(e);
      }
    }
    if (victim != null) {
      final dir = await _sheetsDir();
      for (final name in victim.storedFileNames) {
        final f = File(p.join(dir.path, name));
        if (await f.exists()) {
          await f.delete();
        }
      }
      final parsed = await _parsedFile(victim.id);
      if (await parsed.exists()) {
        await parsed.delete();
      }
    }
    await _writeAll(keep);
  }

  /// 读取结构化结果（不存在或损坏返回 null）。
  Future<SheetParsedData?> loadParsed(String sheetId) async {
    final f = await _parsedFile(sheetId);
    if (!await f.exists()) return null;
    try {
      final decoded = jsonDecode(await f.readAsString());
      return SheetParsedData.fromJson(decoded);
    } catch (_) {
      return null;
    }
  }

  /// 保存结构化结果并回写索引状态。
  Future<void> saveParsed(SheetParsedData data) async {
    final f = await _parsedFile(data.sheetId);
    await f.writeAsString(data.toEncodedJson());
    final all = await loadAll();
    final next = <SheetEntry>[];
    for (final entry in all) {
      if (entry.id != data.sheetId) {
        next.add(entry);
        continue;
      }
      next.add(
        entry.copyWith(
          parseStatus: data.parseStatus.name,
          parsedUpdatedAtMs: data.updatedAtMs,
        ),
      );
    }
    await _writeAll(next);
  }

  /// 解析 [e] 在沙箱中仍存在的页文件（保持 [SheetEntry.storedFileNames] 顺序，跳过缺失）。
  Future<List<File>> resolveStoredFiles(SheetEntry e) async {
    final dir = await _sheetsDir();
    final out = <File>[];
    for (final name in e.storedFileNames) {
      final f = File(p.join(dir.path, name));
      if (await f.exists()) {
        out.add(f);
      }
    }
    return out;
  }

  /// 封面用：第一张仍存在的页；若无则 `null`。
  Future<File?> resolveFirstStoredFile(SheetEntry e) async {
    final list = await resolveStoredFiles(e);
    return list.isEmpty ? null : list.first;
  }
}

File: swift_ios_host/Sources/Sheets/Store/SheetLibraryStore.swift
import Foundation

enum SheetLibraryError: Error, LocalizedError {
    case emptySources
    case invalidDocumentsRoot
    case encodingFailed

    var errorDescription: String? {
        switch self {
        case .emptySources:
            return "sources must not be empty"
        case .invalidDocumentsRoot:
            return "无法访问本地文档目录"
        case .encodingFailed:
            return "编码失败"
        }
    }
}

actor SheetLibraryStore {
    private let indexFileName = "sheets_index.json"
    private let sheetsSubdir = "sheets"
    private let rootOverride: URL?
    private let fileManager: FileManager

    init(rootOverride: URL? = nil, fileManager: FileManager = .default) {
        self.rootOverride = rootOverride
        self.fileManager = fileManager
    }

    func loadAll() async -> [SheetEntry] {
        guard let indexURL = try? indexFileURL() else { return [] }
        guard fileManager.fileExists(atPath: indexURL.path) else { return [] }
        do {
            let data = try Data(contentsOf: indexURL)
            let any = try JSONSerialization.jsonObject(with: data, options: [])
            guard let list = any as? [Any] else { return [] }
            var entries = list.compactMap(SheetEntry.fromJSON)
            entries.sort { $0.addedAtMs > $1.addedAtMs }
            return entries
        } catch {
            return []
        }
    }

    func importSheetPages(sources: [URL], displayName: String) async throws -> SheetEntry {
        if sources.isEmpty { throw SheetLibraryError.emptySources }
        let dir = try sheetsDirURL()
        let id = UUID().uuidString
        var names: [String] = []
        for (idx, src) in sources.enumerated() {
            var ext = src.pathExtension.lowercased()
            if ext.isEmpty { ext = "jpg" }
            let filename = "\(id)_\(idx).\(ext)"
            let dst = dir.appendingPathComponent(filename)
            if fileManager.fileExists(atPath: dst.path) {
                try fileManager.removeItem(at: dst)
            }
            try fileManager.copyItem(at: src, to: dst)
            names.append(filename)
        }

        let entry = SheetEntry(
            id: id,
            displayName: displayName,
            storedFileNames: names,
            addedAtMs: Int(Date().timeIntervalSince1970 * 1000),
            serverPageCount: nil,
            parseStatus: SheetParseStatus.rawOnly.rawValue,
            parsedUpdatedAtMs: nil
        )
        var all = await loadAll()
        all.append(entry)
        try writeAll(entries: all)
        return entry
    }

    func remove(id: String) async throws {
        let all = await loadAll()
        var keep: [SheetEntry] = []
        var victim: SheetEntry?
        for e in all {
            if e.id == id {
                victim = e
            } else {
                keep.append(e)
            }
        }
        if let victim {
            let dir = try sheetsDirURL()
            for name in victim.storedFileNames {
                let file = dir.appendingPathComponent(name)
                if fileManager.fileExists(atPath: file.path) {
                    try? fileManager.removeItem(at: file)
                }
            }
            let parsed = try parsedFileURL(sheetId: victim.id)
            if fileManager.fileExists(atPath: parsed.path) {
                try? fileManager.removeItem(at: parsed)
            }
        }
        try writeAll(entries: keep)
    }

    func loadParsed(sheetId: String) async -> SheetParsedData? {
        guard let f = try? parsedFileURL(sheetId: sheetId) else { return nil }
        guard fileManager.fileExists(atPath: f.path) else { return nil }
        do {
            let data = try Data(contentsOf: f)
            let any = try JSONSerialization.jsonObject(with: data, options: [])
            return SheetParsedData.fromJSON(any)
        } catch {
            return nil
        }
    }

    func saveParsed(_ parsed: SheetParsedData) async throws {
        let f = try parsedFileURL(sheetId: parsed.sheetId)
        let raw = try parsed.toJSONString()
        try raw.write(to: f, atomically: true, encoding: .utf8)

        let all = await loadAll()
        let updated = all.map { e -> SheetEntry in
            guard e.id == parsed.sheetId else { return e }
            var copy = e
            copy.parseStatus = parsed.parseStatus.rawValue
            copy.parsedUpdatedAtMs = parsed.updatedAtMs
            return copy
        }
        try writeAll(entries: updated)
    }

    func resolveStoredFiles(_ entry: SheetEntry) async throws -> [URL] {
        let dir = try sheetsDirURL()
        return entry.storedFileNames
            .map { dir.appendingPathComponent($0) }
            .filter { fileManager.fileExists(atPath: $0.path) }
    }

    func resolveFirstStoredFile(_ entry: SheetEntry) async throws -> URL? {
        let files = try await resolveStoredFiles(entry)
        return files.first
    }

    private func writeAll(entries: [SheetEntry]) throws {
        let arr = entries.map { $0.toJSON() }
        let data = try JSONSerialization.data(withJSONObject: arr, options: [])
        guard let raw = String(data: data, encoding: .utf8) else {
            throw SheetLibraryError.encodingFailed
        }
        try raw.write(to: try indexFileURL(), atomically: true, encoding: .utf8)
    }

    private func docsRoot() throws -> URL {
        if let rootOverride { return rootOverride }
        guard let dir = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            throw SheetLibraryError.invalidDocumentsRoot
        }
        return dir
    }

    private func indexFileURL() throws -> URL {
        try docsRoot().appendingPathComponent(indexFileName)
    }

    private func sheetsDirURL() throws -> URL {
        let dir = try docsRoot().appendingPathComponent(sheetsSubdir, isDirectory: true)
        if !fileManager.fileExists(atPath: dir.path) {
            try fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private func parsedFileURL(sheetId: String) throws -> URL {
        try sheetsDirURL().appendingPathComponent("\(sheetId)_parsed.json")
    }
}
